### DISCUSSION TITLE
Use variable for the hyper-key

### DISCUSSION THREAD

**Discussion by kaspersoerensen** [mxstbr/karabiner#16] [Ideas]
Hi there,

I have been playing around with this for a bit, and find it very useful. I do have some issues with the way the arrow keys are defined: when using the ``V`` layer to navigate, I cannot hold down <kbd>⇧ Shift</kbd> to highlight words. To get around this, i defined a variable called ``hyper`` and used that instead of the modifier keys. That seems to work pretty well ( [see this commit](https://github.com/kaspersoerensen/karabiner/commit/5b56de4ae781f2a3d197321c5eece444075adcac) ).

I also copied the ``createHyperSubLayer`` and ``createHyperSubLayers`` functions and simplified (removed) the part that takes care of the sublayer functionality. This allows me to hold down the <kbd>Caps Lock</kbd> key and use just one key, to execute a command. I use this for the arrow functions. 

I have no experience with TypeScript - so the implementation could be improved (maybe overloading the ``createHyperSubLayer`` and ``createHyperSubLayers`` functions with fewer inputs?).

Maybe this can be useful for someone else?

**Comment by mxstbr** [comment]
Oh damn, the variable hyperkey making Shift to highlight words work is genius! Let me try that as well and see if I can notice any issues. Thanks for the note.

**Reply by mxstbr** [reply]
I tried making this work, but when I copied this implementation none of my existing sublayers worked _and_ I couldn't even type certain characters 🤔

**Reply by kaspersoerensen** [reply]
Wow - that is not what we want. Hmm. Let me test with the ``rules.ts`` you have in your repo.

**Reply by kaspersoerensen** [reply]
Did you update the "Kype Key Definition"? [Check this commit](https://github.com/kaspersoerensen/karabiner/commit/85acb04e62222e97d143609ccec31f8e3d7b4f7c)

**Reply by kaspersoerensen** [reply]
You are right - something in the ``createHyperLayer`` function is not working correctly, so the condition cause is not correctly added to the ``karabiner.json``.

**Reply by mxstbr** [reply]
Figured it out! I made a PR just for the hyperkey variable change to test it out: https://github.com/mxstbr/karabiner/pull/17

**Reply by mxstbr** [reply]
The only thing I haven't been able to figure out yet is using Shift to select text with the movement layer.

**Reply by mxstbr** [reply]
Figured it out! This all works now: https://github.com/mxstbr/karabiner/pull/17

You can also set root hyperkey commands by just adding them to `createHyperSubLayers`, you don't need a separate command for that:

https://github.com/mxstbr/karabiner/pull/17/files#diff-e696162cb54bab305ff3a0d17b701868c09cfc64da63cf2548f63a03dc4c744eR58-R61

**Reply by mxstbr** [reply]
I will test this for today and, if it doesn't break anything, merge it as the main implementation. Thanks for the awesome suggestions!

**Reply by kaspersoerensen** [reply]
Super cool! Thank you so much for implementing this! I look forward to testing this tomorrow!

**Reply by mxstbr** [reply]
Merged #17! Works really well.
